_posts/2025-03-20-phased-strimzi-upgrade-example.md

```diff
@@ -0,0 +1,637 @@
+              3.8.1=quay.io/strimzi/kafka:0.45.0-kafka-3.8.1
+              3.9.0=quay.io/strimzi/kafka:0.45.0-kafka-3.9.0
+        ...
+```
+
+In the sample above you can see that that Kafka version maps to an image which contains the Strimzi version. 
+The container images used by Strimzi include scripts, health endpoint servers and other services which form an API for the container.
+A given version of Strimzi expects certain endpoints, scripts etc to be present on the container in order to perform its operations. 
+This is why, even if the Kafka version has not changed, the image must change when you upgrade Strimzi.
+
+The result of this is that when you upgrade the Cluster Operator it will, as part of the reconciliation, change the image value on **all** of the Kafka broker pods in all of the Kafka clusters managed by that Strimzi instance on the Kubernetes cluster. 
+This means that all of your Kafka fleet (managed by that operator instance) will need to roll. 
+Strimzi will do this one broker at time and maintain topic availability (provided your topics have a replication factor greater than 1).
+However, rolling every Kafka cluster in your fleet could be quite disruptive.
+
+So what can we do to add more control?
+
+## Phased upgrades
+
+What we need is a way to control which Kafka clusters get upgraded and in what order. 
+This is often referred to as a phased upgrade or as an upgrade with _"blast radius"_ control because we can limit the application of the upgrade to a small number of resources (the radius) and contain any potential negative effects (the blast).
+
+There are several possible ways to implement phased upgrades of the Kafka clusters managed by a Strimzi operator:
+
+### Custom image per Kafka cluster
+
+You can prevent the fleet rolling situation by overriding the [image](https://strimzi.io/docs/operators/latest/configuring#con-common-configuration-images-reference) used by each Kafka cluster. 
+You could manually set this to the version used by the current Strimzi Operator version.
+As you have manually set this, the changes to the image map environment variables when you upgrade the Strimzi Operator would not change the images of the operands.
+So no automatic rolling would occur.
+You could then manually switch the image used by each Kafka cluster to the new Strimzi version. 
+This would allow you to control which clusters rolled and when.
+
+However, you would then enter a situation where a new operator version (for example 0.45) is talking to a Kafka cluster running the old Strimzi version's container images (eg 0.44). 
+As mentioned in the introduction, the Strimzi container images represent an API with certain scripts, endpoints and environment variables expected to be present. 
+Running a newer Strimzi version administering a Kafka cluster using older Strimzi images _could_ lead to issues and is not a recommended or supported deployment model.
+
+### Strimzi per Namespace
```

Comment:
It might be worth to emphasize here that only Strimzi versions close to each other can share the CRDs, ClusterRoles etc., so while this is valid way, it requires a lot of self-control to make sure the versions do not diverge too much.